libs/image-editor/ImageEditor/src/main/kotlin/org/wordpress/android/imageeditor/preview/PreviewImageAdapter.kt

```diff
@@ -0,0 +1,41 @@
+package org.wordpress.android.imageeditor.preview
+
+import android.view.ViewGroup
+import android.widget.ImageView
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.ListAdapter
+import androidx.recyclerview.widget.RecyclerView
+import org.wordpress.android.imageeditor.preview.PreviewImageViewModel.ImageData
+import org.wordpress.android.imageeditor.preview.PreviewImageViewModel.ImageUiState
+
+class PreviewImageAdapter(
+    private val loadIntoImageViewWithResultListener: (ImageData, ImageView, Int) -> Unit
+) : ListAdapter<ImageUiState, RecyclerView.ViewHolder>(PreviewImageDiffCallback()) {
+    override fun onCreateViewHolder(
+        parent: ViewGroup,
+        viewType: Int
+    ): RecyclerView.ViewHolder {
+        return PreviewImageViewHolder.create(
+                parent,
+                loadIntoImageViewWithResultListener
+        )
+    }
+
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+        val uiState = getItem(position)
+        (holder as PreviewImageViewHolder).onBind(uiState)
+    }
+}
+
+private class PreviewImageDiffCallback : DiffUtil.ItemCallback<ImageUiState>() {
+    override fun areItemsTheSame(oldItem: ImageUiState, newItem: ImageUiState): Boolean {
+        return oldItem.data.lowResImageUrl == newItem.data.lowResImageUrl
+    }
+
+    override fun areContentsTheSame(oldItem: ImageUiState, newItem: ImageUiState): Boolean {
```

Comment:
I'm wondering if this implementation is what we actually want. AFAIK we want to return `false` when we need to re-render the UI right?

For example I believe we don't need to re-render anything for the following states, right? 
ImageDataStartLoadingUiState
ImageInLowResLoadSuccessUiState
ImageInLowResLoadFailedUiState

SideNote: It raises a question if it make sense to have 3 uiState for something which doesn't result in a change on the UI.

Reply:
I'd consider fixing this. Even though it's just an optimization, it feels like a bug.
Perhaps, something like this?
`
return oldItem.data != newItem.data
           || oldItem.progressBarVisible != newItem.progressBarVisible
           || oldItem.retryLayoutVisible != newItem.retryLayoutVisible;
`

wdyt?

Reply:
>  AFAIK we want to return false when we need to re-render the UI right?

Correct

> For example I believe we don't need to re-render anything for changes between the following states, right?
ImageDataStartLoadingUiState
ImageInLowResLoadSuccessUiState
ImageInLowResLoadFailedUiState

> SideNote: It raises a question if it make sense to have 3 uiState for something which doesn't result in a change on the UI.

I'll think more on it. I just tried to refactor and fit in existing logic and states for single image to multiple images in the first iteration. There might still be a case when we will have just one image in the list and all states might be meaningful, however I haven't thought much about it yet.

Reply:
>I'll think more on it. I just tried to refactor and fit in existing logic and states for single image to multiple images in the first iteration. There might still be a case when we will have just one image in the list and all states might be meaningful, however I haven't thought much about it yet.

I see. Yeah make sense to keep it since we still use it for the single image scenario. I didn't realize that. Thanks for the clarification!